### PR TITLE
Revert chain command

### PR DESCRIPTION
**Pull request by arkpar**
Allows reverting unfinalized blocks for now.

### CODE REVIEW

File: substrate/state-db/src/lib.rs [hunk 1]
```diff
@@ -235,6 +245,20 @@ impl<BlockHash: Hash, Key: Hash> StateDbSync<BlockHash, Key> {
 		}
 	}
 
+	/// Revert all unfinalized blocks with the best block number.
+	/// Returns a database commit or `None` if not possible.
+	/// For archive an empty commit set is returned.
+	pub fn revert_one(&mut self) -> Option<CommitSet<Key>> {
```
Comment (gnunicorn) [inline substrate/state-db/src/lib.rs]: the method name is `_one`, yet the docs say `all`. This is confusing, which one is it?

File: substrate/state-db/src/lib.rs [hunk 2]
```diff
@@ -289,6 +313,19 @@ impl<BlockHash: Hash, Key: Hash> StateDb<BlockHash, Key> {
 	pub fn get<D: HashDb<Hash=Key>>(&self, key: &Key, db: &D) -> Result<Option<DBValue>, Error<D::Error>> {
 		self.db.read().get(key, db)
 	}
+
+	/// Revert all unfinalized blocks with the best block number.
+	/// Returns a database commit or `None` if not possible.
+	/// For archive an empty commit set is returned.
+	pub fn revert_one(&self) -> Option<CommitSet<Key>> {
```
Comment (gnunicorn) [inline substrate/state-db/src/lib.rs]: same as above.

File: polkadot/cli/src/cli.yml [hunk 3]
```diff
@@ -178,3 +178,20 @@ subcommands:
               long: max-heap-pages
               value_name: COUNT
               help: The maximum number of 64KB pages to ever allocate for Wasm execution. Don't alter this unless you know what you're doing.
+  - revert:
+      about: Revert chain to the previous state
+      args:
+          - BLOCKS:
```
Comment (gnunicorn) [inline polkadot/cli/src/cli.yml]: maybe use `BLOCKS COUNT` or `NUM` here? I first thought, you were supposed to provide a list of block hashes or some other way you reference _a block_.

**Comment by arkpar** [comment]
This allows to rewind the chain a few blocks. Useful for debugging and local testing.

File: substrate/state-db/src/lib.rs (hunk 1 above)
Comment (arkpar) [reply substrate/state-db/src/lib.rs]: There may be multiple unfinalized blocks sharing the same number. This reverts a single level of the block tree, including all blocks on that level.

Comment (arkpar) [reply substrate/state-db/src/lib.rs]: Should be read as "revert one block number", but this may include multiple actual blocks.

Comment (gnunicorn) [reply substrate/state-db/src/lib.rs]: Gotcha. The naming is still a bit confusing, however, I don't have any better idea, so go for it.

**Comment by chevdor** [comment]
This feature could also help with issues such as #401, nice.

**Pull request merged**